[AOTI] Embed cubin files into .so

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150739

Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

desertfire committed:
[AOTI] Embed cubin files into .so

Summary: Embed cubin files into .so as an ELF section.

[ghstack-poisoned]

vadimkantorov commented on torch/_inductor/codecache.py:
```diff
@@ -1812,7 +1812,36 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
+                    cmd = [
+                        "ld",
+                        "-r",
+                        "-b",
```
also, as far as I understand, this is specific for gcc-ld :(

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Added an option to fall back to the old behavior of generating individual cubin files when needed.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Added an option to fall back to the old behavior of generating individual cubin files when needed.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Added an option to fall back to the old behavior of generating individual cubin files when needed.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

vadimkantorov commented on torch/_inductor/codecache.py:
```diff
@@ -1812,7 +1812,42 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
+                    obj_file = cubin_file + ".o"
+                    cubins_o.append(obj_file)
+                    # Convert .cubin to .o
+                    cmd = f"{ld} -r -b binary -z noexecstack -o {obj_file} {cubin_file}"
```
Would this syntax work for llvm-ld? Or is this a non-issue?

desertfire replied:
I tried `lld` locally and it worked. Is there any reason you want to use `llvm-ld`?

vadimkantorov replied:
Just worried that on some systems `ld` might be symlinked to `llvm-ld`, and then this might bite. But maybe it's a non-issue

vadimkantorov replied:
E.g. maybe on Apple machines where clang/llvm toolchain is the default...

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

desertfire commented:
@pytorchbot rebase

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

angelayi commented on torch/_inductor/codecache.py:
```diff
@@ -1955,7 +1955,43 @@ def _pad_to_alignment(raw_bytes: bytes) -> bytes:
+                    # By default objcopy will create *_start, *_size, *_end symbols using the full path
+                    # Rename to use the unique kernel name
+                    file_name = re.sub(r"[\W]", "_", cubin_file)
+                    cmd = (
+                        f"{objcopy} "
+                        + f"--redefine-sym _binary_{file_name}_start=__{kernel_name}_start "
+                        + f"--redefine-sym _binary_{file_name}_size=__{kernel_name}_size "
+                        + f"--redefine-sym _binary_{file_name}_end=__{kernel_name}_end "
+                        + obj_file
+                    )
+                    subprocess.run(cmd.split(), capture_output=True, text=True)
```
could this be integrated with the CppBuilder?

desertfire committed:
Update on "[AOTI] Embed cubin files into .so"


Summary: Embed cubin files so AOTI is one step closer to generate a single binary. Controlled by a flag and off as default.

Differential Revision: [D72535357](https://our.internmc.facebook.com/intern/diff/D72535357)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D72535357).

huydhn commented:
@desertfire It looks like one of the test is failing on ROCm inductor/test_aot_inductor_package.py::TestAOTInductorPackageCpp_cuda::test_compile_after_package [GH job link](https://github.com/pytorch/pytorch/actions/runs/15102101998/job/42445355337) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a2d0ef242d20b5b83a28ec2b2fd145bc6e5479e4)

desertfire commented:
> @desertfire It looks like one of the test is failing on ROCm inductor/test_aot_inductor_package.py::TestAOTInductorPackageCpp_cuda::test_compile_after_package [GH job link](https://github.com/pytorch/pytorch/actions/runs/15102101998/job/42445355337) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a2d0ef242d20b5b83a28ec2b2fd145bc6e5479e4)

This PR added an opt-in feature. I think it's ok to skip the test for rocm now. https://github.com/pytorch/pytorch/pull/153828

jeanschmidt commented:
should we revert this PR and merge it back with the one skipping tests?

malfet commented:
> should we revert this PR and merge it back with the one skipping tests?

IMO forward fix is fine, as indeed it's an opt in by design. But please create an issue to track progress of similar feature for ROCm
